src/main/java/uk/nhs/tis/trainee/notifications/service/ProgrammeMembershipService.java

```diff
@@ -356,26 +384,89 @@ public void deleteNotifications(ProgrammeMembership programmeMembership)
+      boolean isDeferral = oldStartDate != null
+          && programmeMembership.getStartDate() != null
+          && oldStartDate.plusDays(DEFERRAL_IF_MORE_THAN_DAYS)
+          .isBefore(programmeMembership.getStartDate());
+      log.info("Programme membership {} is deferral: {} (old start date {}, new start date {})",
+          programmeMembership.getTisId(), isDeferral, oldStartDate,
+          programmeMembership.getStartDate());
+      return isDeferral;
     }
 
-    return true; //immediately notify of a new programme membership
+    return true; //send new notifications
+  }
+
+  /**
+   * Helper function to determine when a deferred notification should be scheduled.
+   *
+   * @param notificationType         The notification type.
+   * @param programmeMembership      The updated programme membership to consider.
+   * @param notificationsAlreadySent The notifications already sent for this entity.
+   * @return the date it should be scheduled, or null if it should be sent immediately.
+   */
+  private Date whenScheduleDeferredNotification(NotificationType notificationType,
+      ProgrammeMembership programmeMembership,
+      Map<NotificationType, History> notificationsAlreadySent) {
+
+    //schedule deferred notifications with the same lead time as the original notification
+    if (notificationsAlreadySent.containsKey(notificationType)) {
+      History lastSent = notificationsAlreadySent.get(notificationType);
+      LocalDate oldStartDate = getProgrammeStartDate(lastSent);
+      LocalDate newStartDate = programmeMembership.getStartDate();
+      if (lastSent.sentAt() != null && oldStartDate != null) {
+        LocalDateTime oldSentDateTime = lastSent.sentAt().atZone(timezone).toLocalDateTime();
+        long leadDays = Duration.between(oldSentDateTime, oldStartDate.atStartOfDay()).toDays();
+        LocalDate newSend = newStartDate.minusDays(leadDays);
+        if (newSend.isAfter(LocalDate.now())) {
+          return Date.from(newSend.atStartOfDay(timezone).toInstant());
+        }
+      }
+      return null; //send immediately if newSend is not in the future, or any data missing
+    }
+
+    return null; //send new notification immediately
+  }
+
+
+  /**
+   * Get the programme start date from a saved history item.
+   *
+   * @param history The history to inspect.
+   * @return The programme start date, or null if it is missing or unparseable.
+   */
+  private LocalDate getProgrammeStartDate(History history) {
+    if (history.template() != null
+        && history.template().variables() != null
+        && history.template().variables().get(START_DATE_FIELD) != null) {
+      //to be deferrable, a programme-related notification must include the START_DATE_FIELD
+      try {
+        return (LocalDate) history.template().variables().get(START_DATE_FIELD);
+      } catch (Exception e) {
+        try {
+          Date startDateAsDate = (Date) history.template().variables().get(START_DATE_FIELD);
```

Comment:
This is a bit clumsy, but the intention is to support both Date and LocalDate formats. The default mapper generates a Date object from what is in mongo.

Reply:
(Possibly all the PM service tests should not be using a LocalDate START_DATE, and rather a Date object; I was getting very confused 🤯 )

Reply:
This feels like we need a custom deserializer rather than having this sort of approach, it's going to trip us up somewhere along the line otherwise.

Reply:
Sorry, I had to force-push with a rebase. The only real change is removing that Date / LocalDate exception handling.